optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/PlannerBenchmarkRunner.java

```diff
@@ -122,54 +127,132 @@ private void initBenchmarkDirectoryAndSubdirs() {
+        int solverBenchmarkResultIndex = 0;
+        for (int i = 0; i < cyclesCount; i++) {
+            long timeCycleEnd = System.currentTimeMillis() + timeLeftPerCycle;
+            for (int j = 0; j < parallelBenchmarkCount; j++) {
+                solverBenchmarkResultCycle[j] = plannerBenchmarkResult.
+                        getSolverBenchmarkResultList().get(solverBenchmarkResultIndex % solverBenchmarkResultCount);
+                solverBenchmarkResultIndex++;
+            }
+            ConcurrentMap<Future<SingleBenchmarkRunner>, SingleBenchmarkRunner> futureMap
+                    = new ConcurrentHashMap<Future<SingleBenchmarkRunner>, SingleBenchmarkRunner>(parallelBenchmarkCount);
+            warmUpPopulate(futureMap, singleBenchmarkResultIndexMap, solverBenchmarkResultCycle, timeLeftPerCycle);
+            warmUp(futureMap, singleBenchmarkResultIndexMap, timeCycleEnd);
+        }
+        restoreTerminationConfigs(originalTerminationConfigMap);
+        List<Runnable> notFinishedWarmUpList = warmUpExecutorService.shutdownNow();
+        if (!notFinishedWarmUpList.isEmpty()) {
+            throw new IllegalStateException("Impossible state: notFinishedWarmUpList (" + notFinishedWarmUpList
+                    + ") is not empty.");
+        }
+        logger.info("================================================================================");
+        logger.info("Warm up ended");
+        logger.info("================================================================================");
+    }
+
+    private void backupTerminationConfigs(Map<SolverBenchmarkResult, TerminationConfig> originalTerminationConfigMap) {
+        for (SolverBenchmarkResult solverBenchmarkResult : plannerBenchmarkResult.getSolverBenchmarkResultList()) {
+            TerminationConfig originalTerminationConfig = solverBenchmarkResult.getSolverConfig().getTerminationConfig();
+            if (!originalTerminationConfigMap.containsKey(solverBenchmarkResult)) {
+                originalTerminationConfigMap.put(solverBenchmarkResult, originalTerminationConfig);
             }
-            logger.info("================================================================================");
-            logger.info("Warm up ended");
-            logger.info("================================================================================");
         }
     }
 
-    protected long warmUp(ProblemBenchmarkResult problemBenchmarkResult,
-            long startingTimeMillis, long warmUpTimeMillisSpentLimit, long timeLeft) {
-        for (SingleBenchmarkResult singleBenchmarkResult : problemBenchmarkResult.getSingleBenchmarkResultList()) {
-            SolverBenchmarkResult solverBenchmarkResult = singleBenchmarkResult.getSolverBenchmarkResult();
+    private void warmUpPopulate(Map<Future<SingleBenchmarkRunner>, SingleBenchmarkRunner> futureMap,
+                                ConcurrentMap<SolverBenchmarkResult, Integer> singleBenchmarkResultIndexMap,
```

Comment:
Code style: method parameters are 2 idents in (so 8 spaces), instead of being aligned to param